Add Topic Statistics Tutorial

dabonnie opened the pull request:
Add tutorial to enable, configure, and view ROS 2 Topic Statistics.

Signed-off-by: Devin Bonnie <dbbonnie@amazon.com>

dabonnie commented:
Note: this depends on the example file added here: https://github.com/ros2/examples/pull/281

dabonnie commented:
@maryaB-osr the content here is ready for review while we wait for https://github.com/ros2/examples/pull/281

maryaB-osr commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+
+    ros2 topic echo /statistics
+
+The terminal should start publishing statistics messages every 10 seconds, unless this duration was
+changed previously as part of the subscription configuration option ``topic_stats_options.publish_period``.
```
On line 133-134, it says that the default is 1s and it looks like you changed to 10s. But here it implies 10s is the default and that's what you'd see if you _didn't_ optionally configure it. Can you clarify?

And, in either case, I think this can be rewritten as so:

```suggestion
The terminal should start publishing statistics messages every 10 seconds, unless the 
``topic_stats_options.publish_period`` subscription configuration was optionally changed earlier in the tutorial.
```

dabonnie replied:
That's a good point. The default is 1 second whereas the example I gave configures it to 10 seconds (to make the statistics data appear interesting).

maryaB-osr replied:
So I think (?) it'd make more sense to say:

> The terminal should start publishing statistics messages every 10 seconds, **because** the 
``topic_stats_options.publish_period`` subscription configuration was optionally changed earlier in the tutorial

dabonnie replied:
Sounds good.

maryaB-osr commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+Here we see the two currently possible calculated statistics for the ``std_msgs::msg::String`` message published
+to ``/topic`` by the ``minimal_publisher``.
+Because the ``std_msgs::msg::String`` does not have a message header, the ``message_age`` calculation cannot be performed,
+so NaNs are provided.
```
Is "returned" a better word than "provided" here?

dabonnie replied:
Agree with returned.

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+.. contents:: Table of Contents
+   :local:
+
+Overview
+--------
+
+This is a short tutorial of how to enable :ref:`Topic Statistics<AboutTopicStats>`
+in ROS 2 and view the published statistics output using command line tools (:ref:`ros2topic<ROS2Topics>`).
```
Will add a short section about Topic Statistics and call out the link to the concepts page.

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+Tasks
+-----
+
+1 Write the subscriber node with statistics enabled node
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Typo, nice catch.

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+1 Write the subscriber node with statistics enabled node
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Download the example talker code by entering the following command:
```
Will do

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+
+.. code-block:: C++
+
+    // configure the collection window and publish period (default 1s)
+    options.topic_stats_options.publish_period = std::chrono::seconds(10);
+
+    // configure the topic name (default '/statistics')
+    // options.topic_stats_options.publish_topic = "/my_topic"
```
That's correct. I left the commented line so it would be easy for someone to provide their own topic name if desired.

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+As in the :ref:`C++ <CppPubSub>` tutorial, we have a subscriber node which receives string messages from the
+``topic`` topic from the ``topic_callback`` function.
+However, what's new is we've added options to configure the subscription to enable topic statistics with
+the ``rclcpp::SubscriptionOptions()``
```
How about options instead of function? The subscription options are a [struct](http://docs.ros2.org/foxy/api/rclcpp/structrclcpp_1_1SubscriptionOptionsBase_1_1TopicStatisticsOptions.html)

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+    // configure the topic name (default '/statistics')
+    // options.topic_stats_options.publish_topic = "/my_topic"
+
+The configurable fields are given as
```
How about: "The configurable fields are described in the following table:"

dabonnie commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,320 @@
+If you changed the ``topic_stats_options.publish_topic`` option then you will see that instead of
+``/statistics``.
+
+The subscriber node you created is publishing statistics, for the topic ``topic``, to the output topic
+``/statistics``.
```
I think that's a good idea.

maryaB-osr commented on source/Tutorials/Topics/Topic-Statistics-Tutorial.rst:
```diff
@@ -0,0 +1,326 @@
+2 Build and run
+^^^^^^^^^^^^^^^
+
+To build, see :ref:`cpppubsub-build-and-run`.
```
```suggestion
To build, see the :ref:`Build and run <cpppubsub-build-and-run>` section in the pub/sub tutorial.
```

dabonnie commented:
> LGTM, just one non-vital change to consider.
> I'll let you merge @dabonnie; not sure if you're planning to put the rqt screenshot, but that can be followed up on later so go ahead and (squash and) merge if you're ready

I just added the screenshot, addressed your comments, and applied a few minor touchups. Could you please take a look again, @maryaB-osr? Thanks!
